### ISSUE TITLE
Make MaintainedModel safe (thread-safe and context-safe)

### ISSUE THREAD

**Issue by hepcat72** [Princeton-LSI-ResearchComputing/tracebase#734]
### Inspiration

The current `MaintainedModel` implementation uses global variables to change the behavior of auto-updates during data loads.  Depending on the apache configuration, those globals could be shared not only among threads, but even between "processes" (think `mod_perl`, where it's all really one process and each request uses the same memory). We currently don't have any issue because we don't have an edit interface.  The only time `.save()` is ever called on any process under apache is during data validation under apache or on the command line during load.  But, if we add an edit interface (even just for admins), if an edit occurs during a user running a data validation, the `save()` behavior could become unpredictable if Apache is configured in a certain way.

While making `MaintainedModel` thread-safe prevents changes in different threads from affecting one another, calls to `disable`/`enable` various modes (like `autoupdates`, `mass_autoupdates`, and `buffering`) can affect serial executions in the *same* thread, deleteriously in the following 2 ways (though all of these such issues were not in our code - these are the reasons controlling modes with globals is bad):

1. If nested methods were to disable/enable blocks of code in the following manner, a portion  of code in the calling method would end up being incorrect: method 1 disables autoupdates, does things, calls method 2 (which disables, does things, then enables, and returns), method 1 does some more things with an unintentionally enabled autoupdate more before finishing.
2. if (as in the case of tests) the modes are not correctly reset at the end of a method because of a caught & handled exception, the mode at the start of the next method would not be in the correct mode.

### Description

Make `MaintainedModel` thread-safe.  I have looked into this and I have a plan.  Right now, it's just in the form of rough notes, which I will put in the comment section below, but the amount of effort will be minimal.  The crux of it is that I will use `thread.local()` to prevent different threads/processes from interfering with one another.

I will also implement a context manager instead of class attributes (which are essentially global variables), to control the modes to mitigate the serial context issues described above, thus it is much safer.

Class attributes will still be used for the context, but that's how I've learned context managers work.

### Alternatives

Tweak the apache config to keep data between threads separate.

### Dependencies

None

### Comment

MaintainedModel improvements over using globals...

- PR #735
  - [x] Eliminate global variables (make them class attributes)
  - [x] Move the setters/getters for the variables into the class as @classmethods
  - [x] Move the decorators into the class as @staticmethods
- PR yet to be submitted
  - [x] use thread.local() to make the class attributes thread-local. This should allow the "global" variables to be "thread-safe"
    - https://docs.python.org/3/library/threading.html#thread-local-data
    - Pay attention to: https://bugs.python.org/issue24020
    - I discovered this strategy via this stack post: https://stackoverflow.com/questions/13171860/how-django-handles-simultaneous-requests-with-concurrency-over-global-variables
    - ~Not sure if `globals()` is a viable alternative.  I don't yet understand what it's for: https://docs.python.org/3/library/functions.html#globals~
- PR yet to be submitted
  - Create 2 more function decorators to simplify things like loading scripts:
    - [x] defer_autoupdates
      - [x] handles the call to perform_buffered_updates (but checks to see if autoupdates are completely disabled first)
      - [x] handles the setting of disable_autoupdates, etc
      - [x] catches exceptions to restore the "global variables" and re-raises
      - [x] restores the "global variables" after function runs
    - [x] no_autoupdates
      - [x] disables both buffering and autoupdates
      - [x] ensures mutually exclusive with defer_autoupdates
      - [x] restores the "global variables" after function runs
- PR yet to be submitted
  - ~[ ] Add a fixture to all tests that clears the buffer and sets "global variables" to default~ Unnecessary. I just need to make completely encapsulated tests for MaintainedModel that are separate from tracebase... not a part of this issue...
    - https://stackoverflow.com/questions/73865462/reset-global-variables-between-two-tests-on-pytest
- ~PR yet to be submitted~ Moved to issue #740
  - ~[ ] Add the ability for setter methods to be checked that they do not use a maintained field to set their value~

-----

### Assumptions

1. List of assumptions that the code will not explicitly address/check
2. E.g. We will assume input is correct (explaining why there is no validation)

### Limitations

1. A list of things this work will specifically not do
2. E.g. This feature will only handle the most frequent use case X

### Affected Components

- change: File path or DB table ...
- add: Environment variable or server setting
- delete: External executable or cron job

### Requirements

- [ ] 1. List of numbered conditions to be met for the feature
- [ ] 2. E.g. Every column/row must display a value, i.e. cannot be empty
- [ ] 3. Numbers for reference & checkboxes for progress tracking

#### Interface Change description

None provided

#### Code Change Description

None provided

#### Tests

- [ ] 1. A description of at least one test for each requirement above.
- [ ] 2. E.g. Test for req 2 that there's an exception when display value is ''
- [ ] 3. Numbers for reference & checkboxes for progress tracking

**Comment by hepcat72** [comment]
I'm currently on vacation, but Zoe is at work and my niece and nephew are still asleep, so I'm doing some "vanity coding".  I want to eventually publish MaintainedModel, so I'm working on this issue to make MaintainedModel more robust and work in other environments.

**Comment by hepcat72** [comment]
Notes, so that I can pick up where I left off in the office...

Add the ability for setter methods to be checked that they do not use a maintained field to set their value.

I still have to:
- [ ] update all the tests
- [ ] update comments and error messages to not reference deleted methods
- [x] Convert the rebuild script
- [x] create a disable_autoupdates decorator and apply it to the validation interface.
  - [x] maybe figure out a way to make this disable defer wrappers it encounters in a sub-context
- [x] I need to check files that DO NOT import MaintainedModel, but make calls to MaintainedModel methods off of the model objects
- [x] I decided that I should obtain the current coordinator anytime I need it.  I decided this when debugging test `test_get_name_triggers_autoupdate`.  It was creating a disabled coordinator in a context manager and then after that context exited, I wanted to call `get_name` off the model object to lazy-update the database, but the popped coordinator retained in the instance was still disabled, so it was not auto-updating.  That was the behavior I'd intended, but when working with the code, if baffled me for a few minutes and I only understood what was happening by issuing some debug prints, so I no longer think that the coordinator should be retained in the MaintainedModel instance.  It should retrieve the current coordinator each time with MaintainedModel.get_coordinator().


Regarding lazy_update, I was looking for a way to do it implicitly, and I noted that I don't want to do it in __init__, because __init__ happens when saving AND querying, so I realized that I need perhaps to find out when a QuerySet iterable is instantiating a Model object.  QuerySet calls Model.from_db() (a class method) when the model is instantiated.  My hope is that models are only instantiated by QuerySet when they are rendered, and that a QuerySet is only used for selects.

- [ ] Override Model.from_db to perform lazy updates implicitly

https://docs.djangoproject.com/en/4.2/ref/models/instances/#django.db.models.Model.from_db

YES - `from_db` is only called as the `QuerySet` iterable iterates, so if I override that method, I should be able to lazy-update implicitly!  See the response to my comment on this answer:

https://stackoverflow.com/a/70521157/2057516
